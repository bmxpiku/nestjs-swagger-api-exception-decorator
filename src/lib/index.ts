import { HttpException } from '@nestjs/common';

import { ApiException } from './decorators/api-exception.decorator';
import {
  ExceptionArguments,
  ExceptionOrExceptionArray,
  ExceptionOrExceptionArrayFunc,
} from './interfaces/api-exception.interface';
import { Options, Template } from './interfaces/options.interface';
import { areExceptionsPassedWithoutArrowFunction } from './utils/exception.util';

/**
 * Build your own custom decorator. This enables you to re-use the same template again without the need to specify it
 * over and over again
 *
 * @param template Any object describing the template which should be shown as example value
 * @param globalOptions Specify the content type
 */
export const buildTemplatedApiExceptionDecorator = <T = Template>(
  template: T,
  globalOptions?: Omit<Options<T>, 'template'>,
) => {
  return <Exception extends HttpException>(exceptions: ExceptionArguments<Exception>, options?: Options<T>) => {
    const mergedOptions = ({ ...globalOptions, template, ...options } as unknown) as Options<Template>;

    if (areExceptionsPassedWithoutArrowFunction(exceptions)) {
      return ApiException(exceptions as ExceptionOrExceptionArray<Exception>, mergedOptions);
    } else {
      return ApiException(exceptions as ExceptionOrExceptionArrayFunc<Exception>, mergedOptions);
    }
  };
};

export { ApiException };
