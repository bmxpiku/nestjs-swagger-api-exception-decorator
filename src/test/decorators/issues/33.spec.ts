import { InternalServerErrorException } from '@nestjs/common';
import { DECORATORS } from '@nestjs/swagger/dist/constants';

import { buildTemplatedApiExceptionDecorator } from '../../../lib';

const TemplatedApiException = buildTemplatedApiExceptionDecorator({
  message: '$message',
  statusCode: '$status',
  errorCode: '$errorCode',
});

class UserAlreadyRegisteredException extends InternalServerErrorException {
  constructor(email: string) {
    super(`${email} already registered`);
  }
}

describe('Issue #33', () => {
  describe('given pre-initialized exception with schema', () => {
    it('should properly use the schema', () => {
      class UserAlreadyRegistered {
        @TemplatedApiException(() => [new UserAlreadyRegisteredException('email@address.com')], {
          enrichSchema: {
            description: 'User already registered',
            externalDocs: {
              url: 'https://someplace.com/doc1',
              description: 'Refer this for more',
            },
          },
        })
        test() {
          return;
        }
      }

      const descriptor = Object.getOwnPropertyDescriptor(UserAlreadyRegistered.prototype, 'test');
      const meta = Reflect.getMetadata(DECORATORS.API_RESPONSE, descriptor.value);

      expect(meta).toMatchSnapshot();
    });
  });
});
