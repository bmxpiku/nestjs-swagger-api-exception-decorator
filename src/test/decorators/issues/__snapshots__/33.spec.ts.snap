// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Issue #33 given pre-initialized exception with schema should properly use the schema 1`] = `
Object {
  "500": Object {
    "content": Object {
      "application/json": Object {
        "examples": Object {
          "UserAlreadyRegisteredException": Object {
            "description": "email@address.com already registered",
            "value": Object {
              "statusCode": 500,
            },
          },
        },
        "schema": Object {
          "description": "User already registered",
          "externalDocs": Object {
            "description": "Refer this for more",
            "url": "https://someplace.com/doc1",
          },
          "properties": Object {
            "statusCode": Object {
              "example": 500,
              "type": "number",
            },
          },
          "required": Array [
            "statusCode",
          ],
          "type": "object",
        },
      },
    },
    "description": "",
    "isArray": undefined,
    "type": undefined,
  },
}
`;
